src/lib/Microsoft.Health.Events/EventHubProcessor/PartitionLockingBackgroundService.cs

```diff
@@ -0,0 +1,487 @@
+                            if (lastModifiedTime < delayThreshold)
+                            {
+                                _logger.LogError(new UnclaimedPartitionException($"Partition {partition} was last claimed at {lastModifiedTime}. The current time is {DateTime.UtcNow}. Delay = {delay.TotalMinutes} minutes"));
+                            }
+                            else
+                            {
+                                _logger.LogTrace($"Partition ownership is up to date for partition {partition}. The last modified time was {lastModifiedTime}. Delay = {delay.TotalMinutes} minutes");
+                            }
+                        }
+                        catch (RequestFailedException ex) when (ex.ErrorCode == BlobErrorCode.BlobNotFound)
+                        {
+                            _logger.LogError(new UnclaimedPartitionException($"Partition {partition} has never been claimed"));
+                            _logger.LogMetric(EventMetrics.PartitionLastClaimedDelay(partition), DateTimeOffset.UtcNow.ToUnixTimeMilliseconds());
+                        }
+                    }
+                }
+            });
+
+            return Task.CompletedTask;
+        }
+
+        private Task RenewOwnership(CancellationToken ct)
+        {
+            Task.Run(async () =>
+            {
+                while (!ct.IsCancellationRequested)
+                {
+                    foreach (var partition in _ownedPartitions)
+                    {
+                        _logger.LogTrace($"Renewing ownership for partition {partition}");
+                        BlobClient ownershipClient = _containerClient.GetBlobClient($"ownedpartitions/{partition}");
+                        var leaseClient = ownershipClient.GetBlobLeaseClient(_processorId);
+                        await leaseClient.RenewAsync(default, ct);
+
+                        // upload to update last modified timestamp
+                        BlobUploadOptions blobUploadOptions = new BlobUploadOptions()
+                        {
+                            Conditions = new BlobRequestConditions()
+                            {
+                                LeaseId = _processorId,
+                            },
+                        };
+                        await ownershipClient.UploadAsync(BinaryData.FromString(string.Empty), blobUploadOptions);
+                    }
+
+                    await Task.Delay(20000, ct);
+                }
+            });
+
+            return Task.CompletedTask;
+        }
+
+        private async Task StartAssignedPartitionProcessor(CancellationToken ct, List<int> ownedPartitions)
+        {
+            var maximumBatchSize = 100;
+
+            string[] partitions = new string[ownedPartitions.Count];
+            for (var i = 0; i < ownedPartitions.Count; i++)
+            {
+                partitions[i] = ownedPartitions.ElementAt(i).ToString();
```

Comment:
ElementAt enumerates the IEnumerable again. Use the indexer or can simplify with a Linq statement

``
string[] partitions = ownedPartitions.Select(i => i.ToString()).ToArray();
``

Reply:
updated